### ISSUE TITLE
Question about Option type guards

### ISSUE THREAD

**Issue by zoontek** [mobily/ts-belt#11]
Hi 👋

A belt port in TS is something I've already dreamed of, so thank you!
I tried adding it to our codebase, but think it might have an issue with the `Option` type guards (`isSome` / `isNone`)

On usage with `ts-pattern` (for pattern maching):

```ts
import { A, O } from "@mobily/ts-belt";
import { match, when } from "ts-pattern";

const array = A.make(Math.floor(Math.random() * 10), "foo");

match(A.get(array, 5))
  .with(when(O.isSome), value => { /* value is O.Option<string>, it should be string */ })
  .with(when(O.isNone), value => { /* value is O.Option<never>, it should be never */ })
  .exhaustive();
```

I checked the types of the given function and it confirms the issue:

```ts
/** Returns `true` if the provided `option` is `Some(value)`, otherwise, returns `false`. */
export declare function isSome<T>(option: Option<T>): option is Option<T>;

/** Returns `true` if the provided `option` is `None`, otherwise, returns `false`. */
export declare function isNone<T>(option: Option<T>): option is Option<never>;
```

Shouldn't it returns `option is T` and `option is never` instead?

Same example with Rescript (`v` is `int`, not `option(int)`):

https://rescript-lang.org/try?code=DYUwLgBAtgngcgV2MAhgI1ANRcBIIC8EAygPZQgAUArAJQBQ9AzgO4CWYAxgBbTxKoMIbLnwBvegB8S5KgDdahAHwQAUkwB0wUgHMATJQBEcnHghsmEAA4AnEExAA7MIYA0EBVIhxSj-ARV1LV0jE1FzS3QHZ0MGAF96IA

**Comment by mobily** [comment]
hello @zoontek! 👋 thank you for posting the issue, but I think the current types are correct, and let me explain:

```rescript
let myNullableValue = Some(5)

switch myNullableValue {
| Some(v) => Js.log2("value is present", v /* ⬅️ int */)
| None => Js.log("value is absent")
}
```

please notice that in your example, `switch` allows extracting `v` from `Some`, so naturally, `v` is `int`

however if you use the `if` statement, like this:

```rescript
if Belt.Option.isSome(myNullableValue) {
  let value = Belt.Option.getExn(myNullableValue /* ⬅️ option<int> */)
  value->ignore
}
```

[link to ReScript try](https://rescript-lang.org/try?code=DYUwLgBAtgngcgV2MAhgI1ANRcBIIC8EAygPZQgAUArAJQBQ9AzgO4CWYAxgBbTxKoMIbLnwBvegB8S5KgDdahAHwQAUkwB0wUgHMATJQBEcnHghsmEAA4AnEExAA7MIYA0EBVIhxSj-ARV1LV0jE1FzS3QHZ0MGAF9GNgAzCAAhEGAwDQB5KzA2Xw0LMgpKWERkdCxTEEUJCAhQSDCzInTMnLyCxw0dcABRAA9HMv5KoRE8BgaWkABaJTYdR1I7ejigA)

`myNullableValue` is still `option<int>` type and you have to extract the value from `Option`, eventually, the same behavior can be found in `ts-belt`

I assume the problem might exist on the `ts-pattern` side, not sure what exactly it can be, I have never used it. Nevertheless, you may want to use other `O` functions instead to extract the value from `Option`: `O.match`, `O.getWithDefault` or `O.getExn`

PS. thanks for all RN libs you created 😍 I use them in every project I work on!

**Comment by zoontek** [comment]
I will check a bit longer to see if I can create `ts-pattern` guard, I don't want to lose pattern matching (I think you understand that 😄)
I close this issue, will publish a comment if I find something.

**Comment by mobily** [comment]
> I don't want to lose pattern matching (I think you understand that 😄)

more than you can imagine 😆 I wish to had pattern matching (a real one) in TS

> will publish a comment if I find something.

I'm looking forward to it!

**Comment by zoontek** [comment]
`ts-pattern` is actually really nice. Sure, not as powerful as Ocaml pattern matching but it does the job correctly.

Here it struggles because Option is implemented using an opaque type (as an abstract class)

**Comment by mobily** [comment]
> it's impossible to create a proper guard to narrow Option<T> to T because Option<T> is not a superset of T

do you have any suggestions on how to tackle this problem? 🤔 I was using `type Option<T> = Some<T> | None` at the very beginning, I don't remember why I did change that tho

**Comment by zoontek** [comment]
Currently having:

```ts
import { A, O } from "@mobily/ts-belt";

const someGuard = <T>(option: O.Option<T>): option is T => O.toNullable(option) != null;

const array = A.make(Math.floor(Math.random() * 10), "foo");
const fifthItem = A.get(array, 5);

if (someGuard(fifthItem)) {
  console.log(fifthItem);
}
```

Will result in (not a proper guard):

<img width="525" alt="Screenshot 2021-12-14 at 11 38 16" src="https://user-images.githubusercontent.com/1902323/145982757-f43e75e6-2e3f-4995-bcb4-e36ec6d534a9.png">

To create proper guards, you can "unwrap" values, like this:

```ts
import { A, O } from "@mobily/ts-belt";
import { match, when } from "ts-pattern";

const someGuard = <T>(option: T | undefined): option is T => option != null;
const noneGuard = <T>(option: T | undefined): option is undefined => option == null;

const array = A.make(Math.floor(Math.random() * 10), "foo");
const fifthItem = A.get(array, 5);

const unwrapped = O.toUndefined(fifthItem);

if (someGuard(unwrapped)) {
  console.log(unwrapped); // string, which is correct!
}
if (noneGuard(unwrapped)) {
  console.log(unwrapped); // undefined, which is correct!
}
```

But it's tedious to use with pattern matching and nested structures:

```ts
// We have a create a new object to "unwrap" option values just to consume it
match(complexObject)
  .with({ aCertainKey: when(someGuard) }, () => {})
  .with({ aCertainKey: when(noneGuard) }, () => {})
  .exhaustive();
```

**Comment by zoontek** [comment]
Unfortunately, it seems that TS pattern patching can't handle generic type guards for now 😞
I guess that you are not interested in loosening the `Option` type to be:

```ts
import { match, not, __ } from "ts-pattern";

type Nullish = null | undefined;
type Option<T> = T | Nullish; // Instead of an opaque type

const __Some = not(__.nullish);
const __None = __.nullish;

const array = A.make(Math.floor(Math.random() * 10), "foo");
const item = A.get(array, 5);

match(item)
  .with(__Some, value => { /* value is string */ })
  .with(__None, value => { /* value is Nullish */ })
  .exhaustive();
```

Which would be logical, as it allows interop with the rest of the TS ecosystem and does not guarantee `ts-belt` usage.
…But could also be logical since this behaviour is possible some modules of `ts-belt` too:

```ts
const array = A.make(Math.floor(Math.random() * 10), "foo");
const item1 = A.get(array, 5); // Option<string>
const item2 = array[5]; // also possible since array is just a string[] and not an opaque type. string | undefined if noUncheckedIndexedAccess is set to true, string otherwise

```

**Comment by mobily** [comment]
@zoontek the `Option` type is now `type Option<T> = T | undefined | null`, this change has been published in `v3.8.0` 🚀 thanks for your all valuable suggestions, I really appreciate that! ❤️
